pkg/sdn/plugin/pod_linux.go

```diff
@@ -252,20 +191,83 @@ func (m *podManager) ipamDel(id string) error {
 	return nil
 }
 
-func isScriptError(err error) bool {
-	_, ok := err.(*exec.ExitError)
-	return ok
+func ensureOvsPort(ovsif *ovs.Interface, hostVeth string) (int, error) {
+	return ovsif.AddPort(hostVeth, -1)
+}
+
+func setupPodFlows(ovsif *ovs.Interface, ofport int, podIP, podMac string, vnid uint32) error {
+	otx := ovsif.NewTransaction()
+
+	// ARP/IP traffic from container
+	otx.AddFlow("table=20, priority=100, in_port=%d, arp, nw_src=%s, arp_sha=%s, actions=load:%d->NXM_NX_REG0[], goto_table:21", ofport, podIP, podMac, vnid)
+	otx.AddFlow("table=20, priority=100, in_port=%d, ip, nw_src=%s, actions=load:%d->NXM_NX_REG0[], goto_table:21", ofport, podIP, vnid)
+
+	// ARP request/response to container (not isolated)
+	otx.AddFlow("table=40, priority=100, arp, nw_dst=%s, actions=output:%d", podIP, ofport)
+
+	// IP traffic to container
+	otx.AddFlow("table=70, priority=100, ip, nw_dst=%s, actions=load:%d->NXM_NX_REG1[], load:%d->NXM_NX_REG2[], goto_table:80", podIP, vnid, ofport)
+
+	return otx.EndTransaction()
 }
 
-// Get the last command (which is prefixed with "+" because of "set -x") and its output
-func getScriptError(output []byte) string {
-	lines := strings.Split(string(output), "\n")
-	for n := len(lines) - 1; n >= 0; n-- {
-		if strings.HasPrefix(lines[n], "+") {
-			return strings.Join(lines[n:], "\n")
+func setupPodBandwidth(ovsif *ovs.Interface, pod *kapi.Pod, hostVeth string) error {
+	podIngress, podEgress, err := kbandwidth.ExtractPodBandwidthResources(pod.Annotations)
+	if err != nil {
+		return fmt.Errorf("failed to parse pod bandwidth: %v", err)
+	}
+	if podIngress == nil && podEgress == nil {
+		return nil
+	}
+
+	var ovsIngress, ovsEgress int64
+	// note pod ingress == OVS egress and vice versa, and OVS ingress is in Kbps
+	if podIngress != nil {
+		ovsEgress = podIngress.Value()
+	}
+	if podEgress != nil {
+		ovsIngress = podEgress.Value() / 1024
+	}
+
+	if ovsEgress > 0 {
+		qos, err := ovsif.Create("qos", "type=linux-htb", fmt.Sprintf("other-config:max-rate=%d", ovsEgress))
```

Comment:
ah, i think this branch originally predates that addition and I must have missed it when rebasing